File: client/src/controller.ts
export enum EventType {
    KeyPress = 1,
    MouseMove,
    MouseClick,
    MouseScroll,
}

export enum MouseButtonType {
    Left = 1,
    Right,
    Middle,
    None,
}

type EventData ={
    EventType: EventType,
    Key: string,
    X: number,
    Y: number,
    MKey: MouseButtonType,
}

class Controller {
    #ws: WebSocket;
    constructor() {
        this.#ws = new WebSocket("ws://localhost:8080/");
    }

    send(data: EventData) {
        this.#ws.send(JSON.stringify(data));
    }
}

export default Controller;

File: server/controller.go
package main

import (
	"log"

	"github.com/go-vgo/robotgo"
)

type EventType uint8

type MouseButtonType uint8

const (
	Left MouseButtonType = iota
	Right
	Middle
)

const (
	KeyPress EventType = iota
	MouseMove
	MouseClick
	MouseScroll
)

func eventResolver(msg *Message) {
	if msg.EventType == KeyPress {
		handleKeyPress(msg)
	} else if msg.EventType == MouseClick {
		handleMouseClick(msg)
	} else if msg.EventType == MouseMove {
		handleMouseMove(msg)
	} else if msg.EventType == MouseScroll {
		handleMouseScroll(msg)
	}
}

func handleKeyPress(msg *Message) {
	log.Println("Handling keypress")
	robotgo.KeyTap(msg.Key)
}

func handleMouseClick(msg *Message) {
	if msg.MKey == Left {
		robotgo.Click()
	} else if msg.MKey == Right {
		robotgo.Click("right")
	} else if msg.MKey == Middle {
		robotgo.Click("wheel")
	}
}

func handleMouseMove(msg *Message) {
	robotgo.MoveSmooth(msg.X, msg.Y)
}

func handleMouseScroll(msg *Message) {

}
